### DISCUSSION TITLE
How to do an insert with a clob column

### DISCUSSION THREAD

**Discussion by JesperSiteimprove** [godror/godror#117] [Q&A] [answered]
Hi

I am trying to do an insert into a table where one of the columns is a clob, but I get the error: 
ORA-01461: can bind a LONG value only for insert into a LONG column

func main() {
	db, err := sql.Open("godror", "xxx/yyy@zzz")
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()
	stmt, err := db.Prepare(`insert into test_clob(id, text) values(1, :text)`)
	if err != nil {
		log.Fatal(err)
	}
	defer stmt.Close()

	stringSlice := make([]string, 0)
	stringSlice = append(stringSlice, strings.Repeat("testing testing", 100*1024))

	binds := make([]interface{}, 0)
	
	binds = append(binds, stringSlice)
	beforeBind := time.Now()
	_, err = stmt.Exec(binds...)
	if err != nil {
		log.Fatal("exec failure: %v\n", err)
	}
	execTime := time.Now().Sub(beforeBind)
	fmt.Printf("rows: %d, exec time in nanoseconds: %d\n", len(stringSlice), execTime.Nanoseconds())
}

I guess that I somehow have to tell godror that the column is of type clob?

**Comment by tgulacsi** [comment]
If it's too long for Oracle, use a godror.Lob:
```
stringSlice := make([]godror.Lob, 0, 1)
stringSlice = append(stringSlice, godror.Lob{IsClob: true, Reader: strings.NewReader(strings.Repeat("testing testing", 100*1024))})
```

**Reply by JesperSiteimprove** [reply]
Thanks for the feedback.

Now I instead get the error:

ORA-22922: nonexistent LOB value

I have also tried to change the sql to an update, and that gives the same error.

Best Regards
Jesper

**Comment by tgulacsi** [comment, marked as answer]
See [TestReadWriteLob](https://github.com/godror/godror/blob/ef87ef72795d1579c68ae653b06bf9daa6738f90/z_test.go#L1158) in `z_test.go` - if still does not work, please post a complete, working example!

**Reply by JesperSiteimprove** [reply]
I finally got it working, though for some reason I have not been able to reproduce the original error with an insert, but only with a merge.

My conclusions.

If you get the "ORA-01461: can bind a LONG value only for insert into a LONG column" error, you must use the godror.Lob

The reason that I got the "ORA-22922: nonexistent LOB value" error trying to use the godror.Lob was that I did the Prepare directly on the sql.DB object and not on a sql.Conn object.
It seems that godror.Lob only works in statements prepared on the sql.Conn, unless you open the sql.DB with `"StandaloneConnection = true"` then it also works on statements prepared on the sql.DB object.

With the original insert statement, it seems that godror.Lob is not necessary, string also works and is actually a lot faster with big strings.

Thanks for the help, I really appreciate it.

Best Regards
Jesper

**Comment by tgulacsi** [comment]
You're right, *sql.DB is a pool, and a *sql.Stmt re-prepares itself on a different connection if the original is in use.
But LOBs need to be on the same connection for its whole life.

So either use an sql.Conn, or an sql.Tx.
